### ISSUE TITLE
Shouldn't `Amount` handle decimals formatting with `{:.8}` ?

### ISSUE THREAD

**Issue by RCasatta** [rust-bitcoin/rust-bitcoin#2136]
I would expect this to work

```
        assert_eq!("0.00000010 BTC", format!("{:.8}", Amount::from_sat(10)));
```

While it fails with:

```
  left: `"0.00000010 BTC"`,
 right: `"0.0000001 BTC"`', bitcoin/src/amount.rs:2379:9
```

**Comment by tcharding** [comment]
Seems like a reasonable expectation.

**Comment by Kixunil** [comment]
Definitely a bug since we intended fully working formatting in `Amount`.

**Comment by yancyribbens** [comment]
I had a quick look at this bug, and it seems like the precision is not respected at all:

```
println!("amount {}", format!("{:.1}", Amount::from_sat(10)));
println!("amount {}", format!("{:.2}", Amount::from_sat(10)));
println!("amount {}", format!("{:.3}", Amount::from_sat(10)));
println!("amount {}", format!("{:.4}", Amount::from_sat(10)));
println!("amount {}", format!("{:.5}", Amount::from_sat(10)));
println!("amount {}", format!("{:.6}", Amount::from_sat(10)));
println!("amount {}", format!("{:.7}", Amount::from_sat(10)));
println!("amount {}", format!("{:.8}", Amount::from_sat(10)));
println!("amount {}", format!("{:.99}", Amount::from_sat(10)));
```

stdout:
```
amount 0.0000001 BTC                                                                           
amount 0.0000001 BTC                                                                           
amount 0.0000001 BTC
amount 0.0000001 BTC
amount 0.0000001 BTC
amount 0.0000001 BTC
amount 0.0000001 BTC
amount 0.0000001 BTC
amount 0.0000001 BTC
```

Also, the tests surrounding this all call macros, which I think could be simplified to helper functions.

**Comment by Kixunil** [comment]
Oh, now I remembered we purposely ignore precision because it'd cut off significant digits which is pretty bad in a financial application. But cutting zeros should be allowed.

**Comment by yancyribbens** [comment]
> Oh, now I remembered we purposely ignore precision because it'd cut off significant digits which is pretty bad in a financial application.

Hmm well maybe panic or something if precision is used, because silently working may be disastrous if there's an assumption the precision is working.

> But cutting zeros should be allowed.

in the bug stated above, it looks like the zero is cut when it should not be.

**Comment by Kixunil** [comment]
Oh, yeah, maybe panic for precision < 8 to be precise.

> in the bug stated above, it looks like the zero is cut when it should not be.

Crap, my mind is broken today, sorry about that!

So yeah - no precision means infinite precision - don't panic, precision 0..8 - panic, precision 8+ - add zeroes.

**Comment by RCasatta** [comment]
> because it'd cut off significant digits which is pretty bad in a financial application

If I ask any spreadsheet to format `0.01` with one decimal  it gives me back `"0.0"`  and it's fine.  

The use case is putting bitcoin amounts in a text table so that you can read numbers in colums. I can get away with just `{:.8}` working but maybe someone has space constraint and wants `{:.3}` the colum total should sum with the underlying sats anyway.

If this is not acceptable, maybe one idea is using the alternate repr for that, eg `{:#.3}` ?

**Comment by apoelstra** [comment]
I agree. `Display` losing data doesn't mean that the data is actually lost (e.g. in a spreadsheet application).

**Comment by Kixunil** [comment]
I was thinking about having `enable_truncating_decimals` method on the builder. What do you think?

**Comment by apoelstra** [comment]
I think that's too confusing and undiscoverable. The precision modifier has an expected behavior (which includes dropping data) and we should just follow that behavior.

**Comment by Kixunil** [comment]
Hmm, I guess you're right.

**Comment by RCasatta** [comment]
#2604 solve the issue as described in the first message.

However, it's not "doing like a spreedsheet" when I ask n digits. 

In other words, the following fails:

```
assert_eq!("0.0 BTC", format!("{:.1}", Amount::from_sat(1)));
```

**Comment by apoelstra** [comment]
Let's re-open the issue. #2604 was by a new contributor who deliberately didn't break any existing tests (which I appreciate) but maybe we need to be more aggressive.

**Comment by 448-OG** [comment]
> #2604 solve the issue as described in the first message.
> 
> However, it's not "doing like a spreedsheet" when I ask n digits.
> 
> In other words, the following fails:
> 
> ```
> assert_eq!("0.0 BTC", format!("{:.1}", Amount::from_sat(1)));
> ```

How would this be handled? Would a new method to handle precision with truncating insignificant zeroes suffice? Maybe something like `Amount::to_precision()`

**Comment by apoelstra** [comment]
No, we don't need a new method. The existing `fmt` method should just handle `precision` correctly.

**Comment by 448-OG** [comment]
Thanks! I would like to solve this. Will follow up for more questions if I don't fully understand the task

**Comment by Kixunil** [comment]
FYI @RCasatta you can workaround this using `amount.display_in(Denomination::Bitcoin).show_denomination()`

**Comment by Kixunil** [comment]
Oh, actually, respecting lower precision is another can of worms. It needs proper rounding.
